GVFS/GVFS.Common/Actions/AbstractObjectCacheAction.cs

```diff
@@ -0,0 +1,184 @@
+                    metadata: null);
+                return false;
+            }
+        }
+
+        public abstract bool IsReady(DateTime now);
+
+        public void Execute()
+        {
+            try
+            {
+                using (FileBasedLock cacheLock = GVFSPlatform.Instance.CreateFileBasedLock(
+                    this.Context.FileSystem,
+                    this.Context.Tracer,
+                    Path.Combine(this.Context.Enlistment.GitObjectsRoot, ObjectCacheLock)))
+                {
+                    if (!cacheLock.TryAcquireLock())
+                    {
+                        this.Context.Tracer.RelatedInfo(this.TelemetryKey + ": Skipping post-fetch work since another process holds the lock");
+                        return;
+                    }
+
+                    lock (this.gitProcessLock)
+                    {
+                        if (this.stopping)
+                        {
+                            return;
+                        }
+
+                        this.gitProcess = new GitProcess(this.Context.Enlistment);
+                    }
+
+                    this.RunGitAction();
+                }
+            }
+            catch (IOException e)
+            {
+                this.Context.Tracer.RelatedWarning(
+                    metadata: this.CreateEventMetadata(e),
+                    message: this.TelemetryKey + ": IOException while running action: " + e.Message,
+                    keywords: Keywords.Telemetry);
+            }
+            catch (Exception e)
+            {
+                this.Context.Tracer.RelatedError(
+                    metadata: this.CreateEventMetadata(e),
+                    message: this.TelemetryKey + ": Exception while running action: " + e.Message,
+                    keywords: Keywords.Telemetry);
+                Environment.Exit((int)ReturnCode.GenericError);
+            }
+        }
+
+        public void Stop()
+        {
+            this.stopping = true;
+            lock (this.gitProcessLock)
+            {
+                GitProcess process = this.gitProcess;
+                if (process != null)
+                {
+                    if (process.TryKillRunningProcess())
+                    {
+                        this.Context.Tracer.RelatedEvent(
+                            EventLevel.Informational,
+                            this.TelemetryKey + ": killed background Git process during " + nameof(this.Stop),
+                            metadata: null);
+                    }
+                    else
+                    {
+                        this.Context.Tracer.RelatedEvent(
+                            EventLevel.Informational,
+                            this.TelemetryKey + ": failed to kill background Git process during " + nameof(this.Stop),
+                            metadata: null);
+                    }
+                }
+            }
+        }
+
+        protected abstract void RunGitAction();

```

Comment:
Comment this method with the preconditions it guarantees (specifically, we hold the object-cache lock before calling this method).

Reply:
(I still need to do this in the new version.)